apps/prairielearn/src/middlewares/authzHasCourseInstanceAccess.js

```diff
@@ -2,9 +2,22 @@ const error = require('@prairielearn/error');
 
 module.exports = function (req, res, next) {
   if (
+    // Effective user is course instructor.
     res.locals.authz_data.has_course_permission_preview ||
+    // Effective user is course instance instructor.
     res.locals.authz_data.has_course_instance_permission_view ||
-    res.locals.authz_data.has_student_access_with_enrollment
+    // Effective user is enrolled in the course instance.
+    res.locals.authz_data.has_student_access_with_enrollment ||
+    // Effective user has student access to the course instance,
+    // and the authenticated user is either a course instructor or
+    // a course instance instructor.
+    //
+    // This is a separate case because we don't automatically enroll
+    // instructors in their own course instances, but we still want
+    // them to be able to access the course instance like a student.
+    (res.locals.authz_data.has_student_access &&
+      (res.locals.authz_data.authn_has_course_permission_preview ||
+        res.locals.authz_data.authn_has_course_instance_permission_view))
```

Comment:
Without this change, the `instructor can request no role and be granted access to student page` test would fail, as the effective user would have no roles and would also not be enrolled.

There might be a more elegant way to handle this case: we could set `has_student_access_with_enrollment = true` in `authzCourseOrInstance` when the instructor is accessing as themselves but without roles. We don't appear to access `has_student_access_with_enrollment` outside of this middleware and `authzCourseOrInstance`, so this should be safe.

Reply:
Sounds good, changed in 109c41c. Please double-check the new conditional!